### PR TITLE
initial add of the Riak Core Connection Manager

### PR DESCRIPTION
**Pull request by bookshelfdave**
This is Chris Tilt's work, I'm just moving files between repos.
see repl PR 212.

cc @buddhisthead @lordnull 

**NOTE** Ranch has been added as a dependency for the connection manager tests to run.

### CODE REVIEW

**Comment by buddhisthead** [comment]
Dave, after more consideration, I believe that we should remove the functions around cluster_finder from the connection manager. It is currently used only by the test modules and I think it is non-used API. We should remove it now, before making it open source, because the connection manager should be devoid of cluster-related connections.

This functionality has been replaced by remote targets, registration of them, etc. The cluster_finder is no longer used. Please cut it all out of the connection_mgr and from the associated test.

File: src/riak_core_service_mgr.erl [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+            case lists:reverse(lists:keysort(1,CommonVers)) of
+                [] ->
+                    %% oops! No common major versions for Proto.
+                    ?TRACE(?debugFmt("Failed to find a common major version for protocol: ~p",
+                                     [ClientProto])),
+                    lager:error("Failed to find a common major version for protocol: ~p", [ClientProto]),
+                    {error,protocol_version_not_supported,Rest};
+                [{Major,CN,HN}] ->
+                    {ok, {ClientProto,Major,CN,HN},Rest};
+                [{Major,CN,HN}, _] ->
```
Comment (lordnull) [inline src/riak_core_service_mgr.erl]: This line should likly be:

```
[{Major,CN,HN} | _] ->
```

As is only 2 version permutations are allowed, anything else will get a case_clause error.

I found this while building a riak_test for cascading repl, which supports 2.0, 1.1, and 1.0 versions.

Comment (buddhisthead) [inline src/riak_core_service_mgr.erl]: Ah! Good catch!
+1 to that change

**Comment by bookshelfdave** [comment]
I added Ranch for the connection manager tests, however it looks like the dep doesn't work with R14B04 | R14B03
https://travis-ci.org/basho/riak_core/builds/5595971

**Comment by buddhisthead** [comment]
Tests Pass.
+1

**Pull request merged**